embedded-graphics/src/primitives/ellipse.rs

```diff
@@ -0,0 +1,459 @@
+pub struct Points {
+    ellipse: Ellipse,
+    iter: super::rectangle::Points,
+}
+
+impl Points {
+    fn new(ellipse: &Ellipse) -> Self {
+        Self {
+            ellipse: *ellipse,
+            iter: ellipse.bounding_box().points(),
+        }
+    }
+
+    fn empty() -> Self {
+        Self {
+            ellipse: Ellipse::new(Point::zero(), Size::zero()),
+            iter: Rectangle::new(Point::zero(), Size::zero()).points(),
+        }
+    }
+}
+
+impl Iterator for Points {
+    type Item = Point;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let ellipse = self.ellipse;
+
+        self.iter.find(|point| ellipse.contains(*point))
+    }
+}
+
+/// Pixel iterator for each pixel in the ellipse border
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    iter: Points,
+    outer_size: Size,
+    outer_color: Option<C>,
+    inner_size: Size,
+    inner_color: Option<C>,
+    center: Point,
+    inner_width: u32,
+}
+
+impl<C> StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    fn new(styled: &Styled<Ellipse, PrimitiveStyle<C>>) -> Self {
+        // Always use a stroke width of 0 if no stroke color was set.
+        let stroke_width = styled.style.effective_stroke_width();
+
+        let outer_size = styled.primitive.size;
+        let inner_size = outer_size.saturating_sub(Size::new(2 * stroke_width, 2 * stroke_width));
+
+        let inner_width = inner_size.width;
+
+        let inner_size = Size::new(inner_size.width.pow(2), inner_size.height.pow(2));
+        let iter = if !styled.style.is_transparent() {
+            Points::new(&styled.primitive)
+        } else {
+            Points::empty()
+        };
+
+        Self {
+            iter,
+            outer_size,
+            outer_color: styled.style.stroke_color,
+            inner_size,
+            inner_color: styled.style.fill_color,
+            center: styled.primitive.center_2x(),
+            inner_width,
+        }
+    }
+}
+
+impl<C> Iterator for StyledEllipseIterator<C>
+where
+    C: PixelColor,
+{
+    type Item = Pixel<C>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self {
+            inner_size,
+            inner_color,
+            outer_color,
+            center,
+            inner_width,
+            ..
+        } = *self;
+
+        self.iter.find_map(|point| {
+            let inner_threshold = size_to_threshold(inner_size, point * 2 - center, inner_width);
+
+            let color = if inner_threshold {
+                inner_color
+            } else {
+                outer_color
+            };
+
+            color.map(|color| Pixel(point, color))
+        })
+    }
+}
+
+impl<'a, C: 'a> Drawable<C> for &Styled<Ellipse, PrimitiveStyle<C>>
+where
+    C: PixelColor,
+{
+    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
+        display.draw_iter(self)
```

Comment:
To be consistent we should probably add `DrawTarget::draw_ellipse`. Do you want to do this in this PR, later or not at all?

Reply:
I've added it but in a separate commit with it's own changelog item. I haven't seen a display that can natively draw an ellipse, but it's here if people want it. I'll rebase this branch into master when it's done to preserve both commits.